pkgs/build-support/gradle/update-deps.nix

```diff
@@ -0,0 +1,154 @@
+      -Dhttps.proxyHost="127.0.0.1"
+      -Dhttp.proxyPort="$MITM_CACHE_PORT"
+      -Dhttps.proxyPort="$MITM_CACHE_PORT"
+      -Djavax.net.ssl.trustStore="$MITM_CACHE_KEYSTORE"
+      -Djavax.net.ssl.trustStorePassword="$MITM_CACHE_KS_PWD"
+    )
+    unset MITM_CACHE_KEYSTORE
+    unset MITM_CACHE_KS_PWD
+    function gradle() {
+      ${gradle}/bin/gradle $gradleFlags "''${gradleFlagsArray[@]}" "''${gradleArgs[@]}" "$@"
+    }
+    export GRADLE_USER_HOME="$(mktemp -d)"
+    export TERM=dumb
+    trap "rm -rf '$GRADLE_USER_HOME'" EXIT
+    export http_proxy="127.0.0.1:$MITM_CACHE_PORT"
+    export https_proxy="127.0.0.1:$MITM_CACHE_PORT"
+    export NIX_SSL_CERT_FILE="$MITM_CACHE_CA"
+    export SSL_CERT_FILE="$MITM_CACHE_CA"
+    SRC="$(mktemp -d)"
+    trap "rm -rf '$SRC'" EXIT
+    cp -r "$source/src" "$SRC/"
+    chmod -R +w "$SRC/src"
+    cd "$SRC/src"
+    source "$source/script2"
+    ${preBuild}
+    ${gradleCommands}
+    ${postBuild}
+  '';
+in
+writeShellScript "fetch-deps.sh" ''
+  set -eo pipefail
+  ${lib.optionalString (lib.hasInfix "$attrPath" code) ''
+    attrPath=${lib.escapeShellArg attrPath}
+    [[ -n "$attrPath" ]] || attrPath="$1"
+    [[ -n "$attrPath" ]] || attrPath="$UPDATE_NIX_ATTR_PATH"
+    [[ -n "$attrPath" ]] || (echo "Set UPDATE_NIX_ATTR_PATH or pass attr path as the first argument!" && exit 1)
+  ''}
+  outPath="${
+    # if this is an absolute path in nix store, use path relative to the output root
+    if lib.hasPrefix "/nix/store/" (toString depsPath)
+    then builtins.concatStringsSep "/" (lib.drop 4 (lib.splitString "/" (toString depsPath)))
+    # if this is an absolute path, just use that path
+    else if lib.hasPrefix "/" (toString depsPath)
+    then depsPath
+    # otherwise, use a path relative to the package
+    else ''$(nix-instantiate --eval -E "dirOf (with (${code}); meta.position)" | tr -d '"')/${depsPath}''
+  }"
+
+  source="$(nix-build --no-out-link -E "import ${./patched-source.nix} ({ self = ${code}; } // builtins.fromJSON "${
+    lib.escapeShellArg (builtins.toJSON (builtins.toJSON attrOverrides))
+  }")")"
+
+  source "$source/script1"
+  pushd "$(mktemp -d)"
+  MITM_CACHE_DIR="$PWD"
+  trap "rm -rf '$MITM_CACHE_DIR'" EXIT
+  ${openssl}/bin/openssl genrsa -out ca.key 2048
+  ${openssl}/bin/openssl req -x509 -new -nodes -key ca.key -sha256 -days 1 -out ca.cer -subj "/C=AL/ST=a/L=a/O=a/OU=a/CN=example.org"
+  export MITM_CACHE_PORT="''${mitmCachePort:-${toString mitmCachePort}}"
+  ${mitm-cache}/bin/mitm-cache \
+    -l"127.0.0.1:$MITM_CACHE_PORT" \
+    record \
+    --reject '\.(md5|sha(1|256|512:?):?)$' \
+    --forget-redirects-from '://(download\.savannah\.gnu\.org|github\.com|packages\.jetbrains\.team|sourceforge\.net:?)/' \
+    --forget-redirects-to '://(cdn\.jsdelivr\.net|plugins-artifacts\.gradle\.org:?)/' \
+    --record-text '/maven-metadata\.xml$' >/dev/null 2>/dev/null &
+  MITM_CACHE_PID="$!"
+  trap "kill '$MITM_CACHE_PID'" EXIT
+  sleep 0.5
```

Comment:
This is not reliable and could cause race conditions on high load machines

Reply:
is there a better way of waiting to make sure the process's successfully started to listen? there's not much message passing possible in bash

Reply:
As a workaround, you could use a named pipe. The process would write EOF to it when ready (non-blocking) and then you could simply  `cat /tmp/waitpipe`.

Alternatively, perhaps mitmproxy could be made to daemonise itself once ready.

Reply:
daemonizing is very annoying when multithreading is involved, so i just used a for loop to check whether the socket is open instead